rust/main/agents/relayer/src/msg/metadata/message_builder.rs

```diff
@@ -0,0 +1,420 @@
+            base,
+            app_context,
+            depth: 0,
+            ism_count: Arc::new(Mutex::new(0)),
+        })
+    }
+
+    pub fn base_builder(&self) -> &Arc<Box<dyn BaseMetadataBuilderTrait>> {
+        &self.base
+    }
+}
+
+/// Builds metadata for a message.
+pub async fn build_message_metadata(
+    message_builder: &MessageMetadataBuilder,
+    message: &HyperlaneMessage,
+    params: MessageMetadataBuildParams,
+) -> Result<IsmWithMetadataAndType> {
+    let ism: Box<dyn InterchainSecurityModule> = message_builder
+        .base_builder()
+        .build_ism(params.ism_address)
+        .await
+        .context("When building ISM")?;
+
+    let module_type = ism
+        .module_type()
+        .await
+        .context("When fetching module type")?;
+
+    // throw error if we've reached max depth or max ism count
+    {
+        let options = &params.options;
+        if message_builder.depth >= options.max_depth {
+            return Ok(IsmWithMetadataAndType {
+                ism,
+                metadata: Metadata::Failed(MetadataBuildError::MaxIsmDepthExceeded(
+                    options.max_depth,
+                )),
+                module_type,
+            });
+        }
+        let mut ism_count = message_builder.ism_count.lock().await;
+        if *ism_count >= options.max_ism_count {
+            return Ok(IsmWithMetadataAndType {
+                ism,
+                metadata: Metadata::Failed(MetadataBuildError::MaxIsmCountReached(
+                    options.max_ism_count,
+                )),
+                module_type,
+            });
+        }
+
+        // update depth and ism count
+        message_builder.depth = message_builder.depth.saturating_add(1);
+        *ism_count = ism_count.saturating_add(1);
+    }
+
+    let metadata: Metadata = match module_type {
+        ModuleType::MerkleRootMultisig => MerkleRootMultisigMetadataBuilder::new(message_builder)
+            .build(message, params)
+            .await
+            .context("When building metadata")?,
+        ModuleType::MessageIdMultisig => MessageIdMultisigMetadataBuilder::new(message_builder)
+            .build(message, params)
+            .await
+            .context("When building metadata")?,
+        ModuleType::Routing => RoutingIsmMetadataBuilder::new(message_builder)
+            .build(message, params)
+            .await
+            .context("When building metadata")?,
+        ModuleType::Aggregation => AggregationIsmMetadataBuilder::new(message_builder)
+            .build(message, params)
+            .await
+            .context("When building metadata")?,
+        ModuleType::CcipRead => CcipReadIsmMetadataBuilder::new(message_builder)
+            .build(message, params)
+            .await
+            .context("When building metadata")?,
+        ModuleType::Null => Metadata::Found(vec![]),
+        _ => return Err(MetadataBuildError::UnsupportedModuleType(module_type).into()),
+    };
```

Comment:
It looks like the original implementation manages to extract method `build` call so that we don't need to repeat it every time. I wonder what may have prevented us to do the same here?

Reply:
Discussed in person. We'll use Box as in original implementation.